src/common/arrstr.cpp

```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
 
     return ret;
 }
+
+#if wxUSE_REGEX
+
+namespace // helpers needed by wxCmpNaturalGeneric()
+{
+// Used for comparison of string parts
+struct wxStringFragment
+{
+    // Fragment types are generally sorted like this:
+    // Empty < SpaceOrPunct < Digit < LetterOrSymbol
+    // Fragments of the same type are compared as follows:
+    // SpaceOrPunct - collated, Digit - as numbers using value
+    // LetterOrSymbol - lower-cased and then collated
+    enum Type
+    {
+        Empty,
+        SpaceOrPunct,  // whitespace or punctuation
+        Digit,         // a sequence of decimal digits
+        LetterOrSymbol // letters and symbols, i.e., anything not covered by the above types
+    };
+
+    wxStringFragment() : type(Empty), value(0) {}
+
+    Type     type;
+    wxString text;
+    wxUint64 value; // used only for Digit type
+};
+
+
+wxStringFragment GetFragment(wxString& text)
+{
+    static const wxRegEx reSpaceOrPunct(wxS("^([[:space:]]|[[:punct:]])+"));
+    static const wxRegEx reDigit(wxS("^[[:digit:]]+"));
+    static const wxRegEx reLetterOrSymbol("^[^[:space:]|[:punct:]|[:digit:]]+");
+
+    wxStringFragment fragment;
+    size_t           length = 0;
+
+    if ( text.empty() )
+        return fragment;
+
+    // In attempt to minimize the number of wxRegEx.Matches() calls,
+    // try to do them from the most expected to the least expected
+    // string fragment type.
+    if ( reLetterOrSymbol.Matches(text) )
+    {
+        if ( reLetterOrSymbol.GetMatch(NULL, &length) )
```

Comment:
In the previous version, there were `wxCHECK()`s but they were all removed, let's see what @vadz has to say here.

Reply:
Sorry, I don't remember all the details about the previous versions, but the principle is always the same: it should never be possible to trigger an assert/check failure due to anything happening outside of the program, i.e. user input, IO error etc. They should only happen in case of preconditions or, more rarely, postconditions failure.

So if we can guarantee that the regex here will always match no matter what, then using `wxCHECK` might make sense. OTOH in this particular case we know that `GetMatch()` will trigger an assertion failure anyhow if it fails (because it can only fail if its preconditions are not satisfied, there is no valid failure mode for it), so it doesn't seem really useful to have it.

Reply:
Actually `GetMatch()` is called only if `Matches()` returns true. So I have no idea what would actually make `GetMatch()` fail here. I have added the if check for `GetMatch()`, it was not in the original PR, perhaps it is useless but I am just to always checking the return value... 

The original `wxCHECK()` here was when `Matches()` for the last possible string fragment type returned false, which should never happen as the regex expression is the negative of the previous two.